clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -2653,20 +2653,34 @@ struct ConvertConstructorToDeductionGuideTransform {
 // Find all template parameters that appear in the given DeducedArgs.
 // Return the indices of the template parameters in the TemplateParams.
 SmallVector<unsigned> TemplateParamsReferencedInTemplateArgumentList(
-    ArrayRef<NamedDecl *> TemplateParams,
+    const TemplateParameterList* TemplateParamsList,
     ArrayRef<TemplateArgument> DeducedArgs) {
   struct TemplateParamsReferencedFinder
       : public RecursiveASTVisitor<TemplateParamsReferencedFinder> {
+    const TemplateParameterList* TemplateParamList;
     llvm::DenseSet<NamedDecl *> TemplateParams;
     llvm::DenseSet<const NamedDecl *> ReferencedTemplateParams;
 
-    TemplateParamsReferencedFinder(ArrayRef<NamedDecl *> TemplateParams)
-        : TemplateParams(TemplateParams.begin(), TemplateParams.end()) {}
+    TemplateParamsReferencedFinder(
+        const TemplateParameterList *TemplateParamList)
+        : TemplateParamList(TemplateParamList),
+          TemplateParams(TemplateParamList->begin(), TemplateParamList->end()) {
+    }
 
     bool VisitTemplateTypeParmType(TemplateTypeParmType *TTP) {
-      MarkAppeared(TTP->getDecl());
+      // We use the index and depth to retrieve the corresponding template
+      // parameter from the parameter list.
+      // Note that Clang may not preserve type sugar during template argument
+      // deduction. In such cases, the TTP is a canonical TemplateTypeParamType,
+      // which only retains its index and depth information.
+      if (TTP->getDepth() == TemplateParamList->getDepth() &&
```

Comment:
Shouldn't this issue come up for the other kind of template parameters as well, besides type parameters?

Reply:
I'm not aware of any issues with non-type parameter cases, I think the AST preserves enough information.

However, for consistency, I have switched the implementation to always use depth and index to retrieve the template parameter declaration, which simplifies the code a bit. Please take a second look.

Reply:
The NTTP case is trickier to test, as we don't canonicalize expressions the same way we do types and templates,
but in some situations, you could have an NTTP which was uniqued pointing to a specific parameter, but this is misleading and you should disregard that and look at only the depth and index.

But I would expect template template parameters to behave similarly as type parameters here, and it should be just as straightforward to test, so I think it would b e worth a try.

Reply:
Thanks. Yeah, template template parameter has similar issues here, added one.